replication/OrderedReplication.tla

```diff
@@ -0,0 +1,349 @@
+                replicaOffsets[r]
+            ]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, inflightReplicationRequest, inflightReplicationResponse, replicaTokens>>
+
+
+(* Recieve request to make a blob permanent.
+ *)
+RecvUpdateTtl(blobId) ==
+    /\ \E r1, r2 \in Replicas:
+        /\ r1 # r2
+        /\ ReplicaCanTakeTtlUpdate(blobId, r1, orderedStoredMessagesPerReplica, replicaStates)
+        /\ orderedStoredMessagesPerReplica' = [r \in DOMAIN orderedStoredMessagesPerReplica |-> 
+            IF (((r = r1) \/ (r = r2)) /\ ReplicaCanTakeTtlUpdate(blobId, r, orderedStoredMessagesPerReplica, replicaStates) /\ ~ReplicaHasUpdateTtlMessage(blobId, r, orderedStoredMessagesPerReplica)) THEN
+                orderedStoredMessagesPerReplica[r] \union { UpdateTtlRecord(blobId, replicaOffsets[r]) }
+            ELSE
+                orderedStoredMessagesPerReplica[r]
+            ]
+        /\ replicaOffsets' = [ r \in DOMAIN replicaOffsets |->
+            IF (((r = r1) \/ (r = r2)) /\ ReplicaCanTakeTtlUpdate(blobId, r, orderedStoredMessagesPerReplica, replicaStates) /\ ~ReplicaHasUpdateTtlMessage(blobId, r, orderedStoredMessagesPerReplica)) THEN 
+                replicaOffsets[r] + 1
+            ELSE 
+                replicaOffsets[r]
+            ]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, inflightReplicationRequest, inflightReplicationResponse, replicaTokens>>
+
+
+(* Ambry uses a pull model for replication.
+ * In SendReplicationRequest action, a replica asks one of its peers to give the set of data the peer has.
+ * This action just enqueues the request to a peer in "inflightReplicationRequest".
+ * The request contains a replicaToken, which is the next offset of the target replica that target replica needs to send as a response.
+ *)
+SendReplicationRequest(src) ==
+    /\ replicaStates[src] = CanTakeWrite \* We want replica asking for the replication pull to be writable.
+    /\ \E tgt \in Replicas:
+        /\ src # tgt
+        /\ orderedStoredMessagesPerReplica[tgt] # {}
+        /\ inflightReplicationRequest' = [inflightReplicationRequest EXCEPT ![tgt] = inflightReplicationRequest[tgt] \union { ReplicationRequestMessage(src, replicaTokens[src][tgt]) } ]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, inflightReplicationResponse, replicaTokens, orderedStoredMessagesPerReplica, replicaOffsets>>
+
+(* Ambry uses a pull model for replication.
+ * In RecvReplicationRequest action, a replica recieves replication request from a peer.
+ * The request contains a replica token, which is the offset of the message upto which the peer is uptodate.
+ * The recieving replica checks if it has message after the offset in the replica token, and sends the message along with the new token.
+ *)
+RecvReplicationRequest(r) ==
+    /\ inflightReplicationRequest[r] # {}
+    /\ orderedStoredMessagesPerReplica[r] # {} \* there should be something to replicate.
```

Comment:
Ok, so the implicit understanding here is that there isn't an active set of records that need to be transmitted to replica r?

Reply:
Oh nevermind.  This is just a check for if the replica is blank.